source/elements/oneDAL/source/data_management/tables.rst

```diff
@@ -1,3 +1,475 @@
+
+      Creates a table object using the implementation passes an a parameter
+
+      :tparam TableImpl: The class that contains the table's implementation
+
+      Invariants
+         | contract ``is_table_impl`` is satisfied
+
+   .. function:: table(const table&)
+
+      Creates new reference object on the table data
+
+   .. function:: table(table&&)
+
+      Moves current reference object into another one
+
+   .. function:: table& operator=(const table&)
+
+      Sets the current object reference to point to another one
+
+   .. member:: std::int64_t feature_count = 0
+
+      The number of :term:`features <Feature>` :math:`p` in the table.
+
+      Getter
+         | ``std::int64_t get_feature_count() const noexcept``
+
+      Invariants
+         | ``feature_count >= 0``
+
+   .. member:: std::int64_t observation_count = 0
+
+      The number of :term:`observations <Observation>` :math:`N` in the table.
+
+      Getter
+         | ``std::int64_t get_observation_count() const noexcept``
+
+      Invariants
+         | ``observation_count >= 0``
+
+   .. member:: bool is_empty = true
+
+      If ``feature_count`` or ``observation_count`` are zero, the
+      table is empty.
+
+      Getter
+         | ``bool is_empty() const noexcept``
+
+   .. member:: table_meta metadata = table_meta()
+
+      The object that represents data structure inside the table
+
+      Getter
+         | ``const dal::table_meta& get_metadata() const noexcept``
+
+      Invariants
+         | ``is_empty = false``
+
+.. _homogen_table:
+
+Homogeneous table
+-----------------
+Class ``homogen_table`` is an implementation of a table type
+for which the following is true:
+
+- Its data is dense and it is stored as one contiguous memory block
+- All features have the same :ref:`data type <Data type>`
+  (but :ref:`feature types <Feature type>` may differ)
+
+::
+
+   class homogen_table : public table {
+   public:
+      // TODO:
+      // Consider constructors with user-provided allocators & deleters
+
+      homogen_table(const homogen_table&);
+      homogen_table(homogen_table&&);
+
+      homogen_table(std::int64_t N, std::int64_t p, data_layout layout);
+
+      template <typename T>
+      homogen_table(const T* const data_pointer, std::int64_t N, std::int64_t p, data_layout layout);
+
+      homogen_table& operator=(const homogen_table&);
+
+      data_type get_data_type() const noexcept;
+      bool has_equal_feature_types() const noexcept;
+
+      template <typename T>
+      const T* get_data_pointer() const noexcept;
+   };
+
+.. namespace:: onedal
+.. class:: homogen_table
+
+   .. function:: homogen_table(const homogen_table&)
+
+      Creates new reference object on the table data
+
+   .. function:: homogen_table(homogen_table&&)
+
+      Moves current reference object into another one
+
+   .. function:: homogen_table(std::int64_t N, std::int64_t p, data_layout layout)
+
+      Creates a homogeneous table with shape of :math:`(N \times p)` with
```

Comment:
repetitions of "with" look clumsy
would "from the default oneDAL allocator" work here?
or, alternatively, "a homogeneous table of shape N x p"?